File: src/Solutions/Euler307.hs
module Solutions.Euler307 where
import Data.List
import Data.Ratio
import Text.Printf

-- binom (10^6, k): choose chips with 2 defects
list1 = let l1 n prev = let a = (prev * (10^6 - n + 1)) `div` n in a:l1 (n + 1) a in take 10001 $ 1:l1 1 1
-- binom (10^6 - k, 20000 - 2k): choose chips with one defect
list2 = let l2 n prev = let t = 10^6 - 10^4 + n
                            a = (prev * t * (t - 2 * n + 1)) `div` ((2 * n - 1) * 2 * n)
                        in a:l2 (n + 1) a
        in reverse $ take 10001 $ 1:l2 1 1
-- binom (20000, 2k): choose defects going to 2-chips
list3 = let l3 n prev = let t = 20000 - 2 * n + 2
                            a = prev * t * (t - 1) `div` (2 * n * (2 * n - 1))
                        in a:l3 (n + 1) a
        in reverse $ take 10001 $ 1:l3 1 1
-- (2k)!/(2!)^k: distribute those defects
list4 = let l4 n prev = let a = (prev * (2 * n - 1) * n) in a:l4 (n + 1) a in take 10001 $ 1:l4 1 1
-- (20000 - 2k)!: distribute remaining defects to 1-chips
list5 = let l5 n prev = let a = prev * (2 * n - 1) * 2 * n in a:l5 (n + 1) a in reverse $ take 10001 $ 1:l5 1 1

s = sum (zipWith5 (\a b c d e -> a * b * c * d * e) list1 list2 list3 list4 list5) :: Integer
total = (10^6)^20000 :: Integer
sol = fromRational (s % total) :: Double

euler307 :: IO String
euler307 = return $ printf "%.10f" $ 1 - sol

File: rust/src/solutions/euler307.rs
use num::{pow::Pow, BigInt};
use std::iter::once;

pub fn euler307() -> String {
    let _1 = || BigInt::from(1);
    // binomial(1_000_000, k): choose chips with 2 defects
    let it1 = once(_1()).chain((1i64..=10000).scan(_1(), |prev, n| {
        let a: BigInt = prev.clone() * (1_000_001 - n) / n;
        *prev = a.clone();
        Some(a)
    }));
    // binomial(1_000_000 - k, 20000 - 2 * k): choose chips with one defect
    let it2 = once(_1())
        .chain(
            (1i64..=10000)
                .scan(_1(), |prev, n| {
                    let t = 1_000_000 - 10_000 + n;
                    let a: BigInt = prev.clone() * t * (t + 1 - 2 * n) / ((2 * n - 1) * 2 * n);
                    *prev = a.clone();
                    Some(a)
                })
                .collect::<Vec<_>>(),
        )
        .rev();
    // binomial(20_000, 2 * k): choose defects going to 2-chips
    let it3 = once(_1())
        .chain(
            (1i64..=10000)
                .scan(_1(), |prev, n| {
                    let t = 20_002 - 2 * n;
                    let a: BigInt = prev.clone() * t * (t - 1) / (2 * n * (2 * n - 1));
                    *prev = a.clone();
                    Some(a)
                })
                .collect::<Vec<_>>(),
        )
        .rev();
    // (2 * k)! / 2^k: distribute those defects
    let it4 = once(_1()).chain((1i64..=10000).scan(_1(), |prev, n| {
        let a: BigInt = prev.clone() * (2 * n - 1) * n;
        *prev = a.clone();
        Some(a)
    }));
    // 20_000 - 2 * k)!: distribute remaining defects to 1-chips
    let it5 = once(_1())
        .chain(
            (1i64..=10000)
                .scan(_1(), |prev, n| {
                    let a: BigInt = prev.clone() * (2 * n - 1) * 2 * n;
                    *prev = a.clone();
                    Some(a)
                })
                .collect::<Vec<_>>(),
        )
        .rev();

    let s = it1
        .zip(it2)
        .map(|(a, b)| a * b)
        .zip(it3)
        .map(|(a, b)| a * b)
        .zip(it4)
        .map(|(a, b)| a * b)
        .zip(it5)
        .map(|(a, b)| a * b)
        .sum::<BigInt>();
    let total = BigInt::from(1_000_000).pow(20_000u32);
    format!("0.{}", &(total - s).to_string()[0..10]) // Nasty
}
